### ISSUE TITLE
Impossible reorg, please file an issue

### ISSUE THREAD

**Issue by ligi** [ethereum/go-ethereum#19302]
filing an issue as demanded by the log. Happening with my goerli node.

```
ligi@ligi-w520:~$ geth --datadir .goerli/ --goerli 
INFO [03-19|18:25:31.721] Maximum peer count                       ETH=25 LES=0 total=25
INFO [03-19|18:25:31.723] Starting peer-to-peer node               instance=Geth/v1.9.0-unstable-3ee09ba0/linux-amd64/go1.11.5
INFO [03-19|18:25:31.723] Allocated trie memory caches             clean=256.00MiB dirty=256.00MiB
INFO [03-19|18:25:31.723] Allocated cache and file handles         database=/home/ligi/.goerli/geth/chaindata cache=512.00MiB handles=524288
INFO [03-19|18:25:31.793] Persisted trie from memory database      nodes=361 size=51.44KiB time=1.461151ms gcnodes=0 gcsize=0.00B gctime=0s livenodes=1 livesize=0.00B
INFO [03-19|18:25:31.793] Initialised chain configuration          config="{ChainID: 5 Homestead: 0 DAO: <nil> DAOSupport: true EIP150: 0 EIP155: 0 EIP158: 0 Byzantium: 0 Constantinople: 0  ConstantinopleFix: 0 Engine: clique}"
INFO [03-19|18:25:31.794] Initialising Ethereum protocol           versions="[63 62]" network=5
INFO [03-19|18:25:31.866] Loaded most recent local header          number=229061 hash=9e9903…ba17d0 td=387285 age=6d19h28m
INFO [03-19|18:25:31.866] Loaded most recent local full block      number=229061 hash=9e9903…ba17d0 td=387285 age=6d19h28m
INFO [03-19|18:25:31.867] Loaded most recent local fast block      number=229061 hash=9e9903…ba17d0 td=387285 age=6d19h28m
INFO [03-19|18:25:31.867] Loaded local transaction journal         transactions=0 dropped=0
INFO [03-19|18:25:31.867] Regenerated local transaction journal    transactions=0 accounts=0
WARN [03-19|18:25:31.867] Blockchain not empty, fast sync disabled 
INFO [03-19|18:25:32.114] New local node record                    seq=71 id=7daf92cf1dd47739 ip=127.0.0.1 udp=30303 tcp=30303
INFO [03-19|18:25:32.114] Started P2P networking                   self=enode://37cf5cff7438500ed2917b18e7c6968fdb4c6beaea4321ef07a6ac794f835419f1e52794d4ef5e6ecba648d1167841ed88dfeff4ba114c21e61cc043309902d7@127.0.0.1:30303
INFO [03-19|18:25:32.117] IPC endpoint opened                      url=/home/ligi/.goerli/geth.ipc
INFO [03-19|18:25:34.097] Mapped network port                      proto=tcp extport=30303 intport=30303 interface="UPNP IGDv1-IP1"
INFO [03-19|18:25:34.171] Mapped network port                      proto=udp extport=30303 intport=30303 interface="UPNP IGDv1-IP1"
INFO [03-19|18:25:34.308] New local node record                    seq=72 id=7daf92cf1dd47739 ip=178.19.221.38 udp=30303 tcp=30303
INFO [03-19|18:25:38.445] Block synchronisation started 
INFO [03-19|18:25:41.129] Importing heavy sidechain segment        blocks=2048 start=229087 end=231134
ERROR[03-19|18:25:41.141] Impossible reorg, please file an issue   oldnum=229061 oldhash=9e9903…ba17d0 newnum=229061 newhash=9e9903…ba17d0
WARN [03-19|18:25:41.143] Synchronisation failed, dropping peer    peer=46b0192be704996d err="retrieved hash chain is invalid"
INFO [03-19|18:25:45.167] Importing heavy sidechain segment        blocks=2048 start=229239 end=231286
ERROR[03-19|18:25:45.217] Impossible reorg, please file an issue   oldnum=229087 oldhash=8359ff…c60b27 newnum=229087 newhash=8359ff…c60b27
WARN [03-19|18:25:45.229] Synchronisation failed, dropping peer    peer=edcb409cc029ffdd err="retrieved hash chain is invalid"
^CINFO [03-19|18:25:46.902] Got interrupt, shutting down... 
INFO [03-19|18:25:46.903] IPC endpoint closed                      url=/home/ligi/.goerli/geth.ipc
INFO [03-19|18:25:46.903] Writing cached state to disk             block=229239 hash=ef5a94…b4e6ee root=a72241…c40a0c
INFO [03-19|18:25:46.904] Persisted trie from memory database      nodes=23  size=7.14KiB  time=868.931µs  gcnodes=1 gcsize=564.00B gctime=4.642µs livenodes=1 livesize=0.00B
INFO [03-19|18:25:46.904] Writing cached state to disk             block=229238 hash=bfd031…e33084 root=8b54e2…e4c5eb
INFO [03-19|18:25:46.904] Persisted trie from memory database      nodes=0   size=0.00B    time=6.105µs    gcnodes=0 gcsize=0.00B   gctime=0s      livenodes=1 livesize=0.00B
INFO [03-19|18:25:46.904] Writing cached state to disk             block=229112 hash=54e578…672a90 root=8b54e2…e4c5eb
INFO [03-19|18:25:46.904] Persisted trie from memory database      nodes=0   size=0.00B    time=4.099µs    gcnodes=0 gcsize=0.00B   gctime=0s      livenodes=1 livesize=0.00B
INFO [03-19|18:25:46.904] Blockchain manager stopped 
INFO [03-19|18:25:46.904] Stopping Ethereum protocol 
INFO [03-19|18:25:46.905] Ethereum protocol stopped 
INFO [03-19|18:25:46.905] Transaction pool stopped 
INFO [03-19|18:25:46.908] Database closed                          database=/home/ligi/.goerli/geth/chaindata
```

maybe related: #19258 #16641

bit more log:
```

INFO [03-19|18:21:09.206] Maximum peer count                       ETH=25 LES=0 total=25
INFO [03-19|18:21:09.208] Starting peer-to-peer node               instance=Geth/v1.9.0-unstable-3ee09ba0/linux-amd64/go1.11.5
INFO [03-19|18:21:09.208] Allocated trie memory caches             clean=256.00MiB dirty=256.00MiB
INFO [03-19|18:21:09.208] Allocated cache and file handles         database=/home/ligi/.goerli/geth/chaindata cache=512.00MiB handles=524288
INFO [03-19|18:21:10.863] Persisted trie from memory database      nodes=361 size=51.44KiB time=1.01315ms gcnodes=0 gcsize=0.00B gctime=0s livenodes=1 livesize=0.00B
INFO [03-19|18:21:10.863] Initialised chain configuration          config="{ChainID: 5 Homestead: 0 DAO: <nil> DAOSupport: true EIP150: 0 EIP155: 0 EIP158: 0 Byzantium: 0 Constantinople: 0  ConstantinopleFix: 0 Engine: clique}"
INFO [03-19|18:21:10.864] Initialising Ethereum protocol           versions="[63 62]" network=5
INFO [03-19|18:21:10.890] Loaded most recent local header          number=228822 hash=2ad016…8ebe94 td=386883 age=6d20h24m
INFO [03-19|18:21:10.890] Loaded most recent local full block      number=228822 hash=2ad016…8ebe94 td=386883 age=6d20h24m
INFO [03-19|18:21:10.890] Loaded most recent local fast block      number=228822 hash=2ad016…8ebe94 td=386883 age=6d20h24m
INFO [03-19|18:21:10.890] Loaded local transaction journal         transactions=0 dropped=0
INFO [03-19|18:21:10.890] Upgrading chain index                    type=bloombits percentage=89
INFO [03-19|18:21:10.890] Regenerated local transaction journal    transactions=0 accounts=0
WARN [03-19|18:21:10.890] Blockchain not empty, fast sync disabled
INFO [03-19|18:21:11.129] New local node record                    seq=69 id=7daf92cf1dd47739 ip=127.0.0.1 udp=30303 tcp=30303
INFO [03-19|18:21:11.130] Stats daemon started
INFO [03-19|18:21:11.130] Started P2P networking                   self=enode://37cf5cff7438500ed2917b18e7c6968fdb4c6beaea4321ef07a6ac794f835419f1e52794d4ef5e6ecba648d1167841ed88dfeff4ba114c21e61cc043309902d7@127.0.0.1:30303
INFO [03-19|18:21:11.132] IPC endpoint opened                      url=/home/ligi/.goerli/geth.ipc
INFO [03-19|18:21:12.568] Finished upgrading chain index           type=bloombits
INFO [03-19|18:21:13.229] Mapped network port                      proto=udp extport=30303 intport=30303 interface="UPNP IGDv1-IP1"
INFO [03-19|18:21:13.301] Mapped network port                      proto=tcp extport=30303 intport=30303 interface="UPNP IGDv1-IP1"
INFO [03-19|18:21:15.133] New local node record                    seq=70 id=7daf92cf1dd47739 ip=178.19.221.38 udp=30303 tcp=30303
INFO [03-19|18:21:19.401] Block synchronisation started
WARN [03-19|18:21:20.730] Failed to decode stats server message    err="json: cannot unmarshal string into Go value of type map[string][]interface {}"
INFO [03-19|18:21:24.852] Importing heavy sidechain segment        blocks=2048 start=228827 end=230874
ERROR[03-19|18:21:24.861] Impossible reorg, please file an issue   oldnum=228822 oldhash=2ad016…8ebe94 newnum=228822 newhash=2ad016…8ebe94
WARN [03-19|18:21:24.874] Synchronisation failed, dropping peer    peer=9573268ff1373203 err="retrieved hash chain is invalid"
WARN [03-19|18:21:26.339] Full stats report failed                 err="tls: use of closed connection"
[[INFO [03-19|18:21:45.454] Importing heavy sidechain segment        blocks=2048 start=228834 end=230881
ERROR[03-19|18:21:45.464] Impossible reorg, please file an issue   oldnum=228831 oldhash=42039e…af75f4 newnum=228831 newhash=42039e…af75f4
WARN [03-19|18:21:45.465] Synchronisation failed, dropping peer    peer=c709d2438b1c53a9 err="retrieved hash chain is invalid"
WARN [03-19|18:21:50.731] Failed to decode stats server message    err="json: cannot unmarshal string into Go value of type map[string][]interface {}"
INFO [03-19|18:21:52.506] Importing heavy sidechain segment        blocks=2048 start=228857 end=230904
ERROR[03-19|18:21:52.514] Impossible reorg, please file an issue   oldnum=228835 oldhash=a436b7…e52069 newnum=228835 newhash=a436b7…e52069
WARN [03-19|18:21:52.517] Synchronisation failed, dropping peer    peer=9573268ff1373203 err="retrieved hash chain is invalid"
WARN [03-19|18:21:56.497] Full stats report failed                 err="tls: use of closed connection"
INFO [03-19|18:22:07.067] Importing heavy sidechain segment        blocks=2048 start=228888 end=230935
ERROR[03-19|18:22:07.075] Impossible reorg, please file an issue   oldnum=228857 oldhash=4d57cd…c9f823 newnum=228857 newhash=4d57cd…c9f823
WARN [03-19|18:22:07.079] Synchronisation failed, dropping peer    peer=29b11550591a1949 err="retrieved hash chain is invalid"
WARN [03-19|18:22:20.730] Failed to decode stats server message    err="json: cannot unmarshal string into Go value of type map[string][]interface {}"
INFO [03-19|18:22:21.678] Importing heavy sidechain segment        blocks=2048 start=228908 end=230955
ERROR[03-19|18:22:21.689] Impossible reorg, please file an issue   oldnum=228889 oldhash=e8b0e6…d167e5 newnum=228889 newhash=e8b0e6…d167e5
WARN [03-19|18:22:21.691] Synchronisation failed, dropping peer    peer=6c4bc6996be0a461 err="retrieved hash chain is invalid"
WARN [03-19|18:22:26.652] Full stats report failed                 err="tls: use of closed connection"
INFO [03-19|18:22:39.167] Importing heavy sidechain segment        blocks=2048 start=228912 end=230959
ERROR[03-19|18:22:39.176] Impossible reorg, please file an issue   oldnum=228908 oldhash=ac232a…76c220 newnum=228908 newhash=ac232a…76c220
WARN [03-19|18:22:39.177] Synchronisation failed, dropping peer    peer=6362a752b7aeed0b err="retrieved hash chain is invalid"
INFO [03-19|18:22:48.934] Importing heavy sidechain segment        blocks=2048 start=228917 end=230964
ERROR[03-19|18:22:48.941] Impossible reorg, please file an issue   oldnum=228912 oldhash=1543cd…85b832 newnum=228912 newhash=1543cd…85b832
WARN [03-19|18:22:48.959] Synchronisation failed, dropping peer    peer=9573268ff1373203 err="retrieved hash chain is invalid"
WARN [03-19|18:22:50.731] Failed to decode stats server message    err="json: cannot unmarshal string into Go value of type map[string][]interface {}"
WARN [03-19|18:22:56.809] Full stats report failed                 err="tls: use of closed connection"
INFO [03-19|18:23:06.695] Importing heavy sidechain segment        blocks=2048 start=228922 end=230969
ERROR[03-19|18:23:06.714] Impossible reorg, please file an issue   oldnum=228919 oldhash=37604e…7a7c4c newnum=228919 newhash=37604e…7a7c4c
WARN [03-19|18:23:06.715] Synchronisation failed, dropping peer    peer=29b11550591a1949 err="retrieved hash chain is invalid"
INFO [03-19|18:23:18.077] Importing heavy sidechain segment        blocks=2048 start=228966 end=231013
ERROR[03-19|18:23:18.084] Impossible reorg, please file an issue   oldnum=228922 oldhash=80e657…0dd0a0 newnum=228922 newhash=80e657…0dd0a0
WARN [03-19|18:23:18.087] Synchronisation failed, dropping peer    peer=6362a752b7aeed0b err="retrieved hash chain is invalid"
WARN [03-19|18:23:20.731] Failed to decode stats server message    err="json: cannot unmarshal string into Go value of type map[string][]interface {}"
WARN [03-19|18:23:26.963] Full stats report failed                 err="tls: use of closed connection"
INFO [03-19|18:23:38.410] Importing heavy sidechain segment        blocks=2048 start=229001 end=231048
ERROR[03-19|18:23:38.419] Impossible reorg, please file an issue   oldnum=228967 oldhash=58f9a2…a51fd6 newnum=228967 newhash=58f9a2…a51fd6
WARN [03-19|18:23:38.421] Synchronisation failed, dropping peer    peer=e28e855ceaa73eaa err="retrieved hash chain is invalid"

```

**Comment by jwasinger** [comment]
I'm currently on mobile so I can't look at this atm.  Can you tell me if the problem persists when you disable fast sync?

**Comment by ligi** [comment]
seeing the same when starting with syncmode full
```
ligi@ligi-w520:~$ geth --datadir .goerli/ --goerli --syncmode "full"
INFO [03-19|18:48:57.453] Maximum peer count                       ETH=25 LES=0 total=25
INFO [03-19|18:48:57.456] Starting peer-to-peer node               instance=Geth/v1.9.0-unstable-3ee09ba0/linux-amd64/go1.11.5
INFO [03-19|18:48:57.456] Allocated trie memory caches             clean=256.00MiB dirty=256.00MiB
INFO [03-19|18:48:57.456] Allocated cache and file handles         database=/home/ligi/.goerli/geth/chaindata cache=512.00MiB handles=524288
INFO [03-19|18:48:57.492] Persisted trie from memory database      nodes=361 size=51.44KiB time=1.38014ms gcnodes=0 gcsize=0.00B gctime=0s livenodes=1 livesize=0.00B
INFO [03-19|18:48:57.492] Initialised chain configuration          config="{ChainID: 5 Homestead: 0 DAO: <nil> DAOSupport: true EIP150: 0 EIP155: 0 EIP158: 0 Byzantium: 0 Constantinople: 0
ConstantinopleFix: 0 Engine: clique}"
INFO [03-19|18:48:57.493] Initialising Ethereum protocol           versions="[63 62]" network=5
INFO [03-19|18:48:57.541] Loaded most recent local header          number=229460 hash=979082…51a3d6 td=387982 age=6d18h12m
INFO [03-19|18:48:57.541] Loaded most recent local full block      number=229460 hash=979082…51a3d6 td=387982 age=6d18h12m
INFO [03-19|18:48:57.541] Loaded most recent local fast block      number=229460 hash=979082…51a3d6 td=387982 age=6d18h12m
INFO [03-19|18:48:57.541] Loaded local transaction journal         transactions=0 dropped=0
INFO [03-19|18:48:57.541] Regenerated local transaction journal    transactions=0 accounts=0
INFO [03-19|18:48:57.652] New local node record                    seq=77 id=7daf92cf1dd47739 ip=127.0.0.1 udp=30303 tcp=30303
INFO [03-19|18:48:57.652] Started P2P networking                   self=enode://37cf5cff7438500ed2917b18e7c6968fdb4c6beaea4321ef07a6ac794f835419f1e52794d4ef5e6ecba648d1167841ed88dfeff4ba114c21e61cc043309902d7@127.0.0.1:30303
INFO [03-19|18:48:57.654] IPC endpoint opened                      url=/home/ligi/.goerli/geth.ipc
INFO [03-19|18:48:59.772] Mapped network port                      proto=udp extport=30303 intport=30303 interface="UPNP IGDv1-IP1"
INFO [03-19|18:48:59.846] Mapped network port                      proto=tcp extport=30303 intport=30303 interface="UPNP IGDv1-IP1"
INFO [03-19|18:49:01.656] New local node record                    seq=78 id=7daf92cf1dd47739 ip=178.19.221.38 udp=30303 tcp=30303
INFO [03-19|18:49:07.547] Block synchronisation started
[INFO [03-19|18:49:09.549] Importing heavy sidechain segment        blocks=2048 start=229473 end=231520
ERROR[03-19|18:49:09.562] Impossible reorg, please file an issue   oldnum=229460 oldhash=979082…51a3d6 newnum=229460 newhash=979082…51a3d6
WARN [03-19|18:49:09.566] Synchronisation failed, dropping peer    peer=6362a752b7aeed0b err="retrieved hash chain is invalid"
INFO [03-19|18:49:13.568] Importing heavy sidechain segment        blocks=2048 start=229476 end=231523
ERROR[03-19|18:49:13.583] Impossible reorg, please file an issue   oldnum=229474 oldhash=3931ad…bd68a5 newnum=229474 newhash=3931ad…bd68a5
WARN [03-19|18:49:13.595] Synchronisation failed, dropping peer    peer=9573268ff1373203 err="retrieved hash chain is invalid"
INFO [03-19|18:49:23.369] Importing heavy sidechain segment        blocks=2048 start=229563 end=231610
ERROR[03-19|18:49:23.380] Impossible reorg, please file an issue   oldnum=229478 oldhash=be712a…b3421a newnum=229478 newhash=be712a…b3421a
WARN [03-19|18:49:23.385] Synchronisation failed, dropping peer    peer=1ef2c5739c2a3734 err="retrieved hash chain is invalid"
^[[A^[[A^[[A^[[A^[[A^[[A^[[A^[[A^[[A^[[A^[[A^[[A^[[A^[[A^[[AINFO [03-19|18:49:34.157] Importing heavy sidechain segment        blocks=2048 start=229578 end=231625
ERROR[03-19|18:49:34.166] Impossible reorg, please file an issue   oldnum=229563 oldhash=762164…326c10 newnum=229563 newhash=762164…326c10
WARN [03-19|18:49:34.167] Synchronisation failed, dropping peer    peer=6362a752b7aeed0b err="retrieved hash chain is invalid"
INFO [03-19|18:49:46.654] Importing heavy sidechain segment        blocks=2048 start=229688 end=231735
ERROR[03-19|18:49:46.691] Impossible reorg, please file an issue   oldnum=229578 oldhash=9ee68c…59b352 newnum=229578 newhash=9ee68c…59b352
WARN [03-19|18:49:46.702] Synchronisation failed, dropping peer    peer=46b0192be704996d err="retrieved hash chain is invalid"
INFO [03-19|18:50:06.501] Importing heavy sidechain segment        blocks=2048 start=229696 end=231743
ERROR[03-19|18:50:06.512] Impossible reorg, please file an issue   oldnum=229689 oldhash=75fab6…2453fe newnum=229689 newhash=75fab6…2453fe
WARN [03-19|18:50:06.513] Synchronisation failed, dropping peer    peer=6362a752b7aeed0b err="retrieved hash chain is invalid"
INFO [03-19|18:50:23.883] Importing heavy sidechain segment        blocks=2048 start=229845 end=231892
ERROR[03-19|18:50:23.927] Impossible reorg, please file an issue   oldnum=229696 oldhash=e4a6c8…05366e newnum=229696 newhash=e4a6c8…05366e
WARN [03-19|18:50:23.934] Synchronisation failed, dropping peer    peer=b9f7252e41a9d26c err="retrieved hash chain is invalid"

```

**Comment by holiman** [comment]
@ligi did you by any chance exit unclean earlier, e.g. due to out of disk?

**Comment by jwasinger** [comment]
`oldnum=229696 oldhash=e4a6c8…05366e newnum=229696 newhash=e4a6c8…05366e`

So this looks like the edge case where two sealers generate blocks at the exact same time is being reached here.  I'm taking a shot at creating a patch for this problem.

**Comment by ligi** [comment]
> @ligi did you by any chance exit unclean earlier, e.g. due to out of disk?

yes - actually there was such an incident (spontaneous shutdown of the machine) - did not correlate the 2 events as this node is starting in the background and I did not check after the incident - but they could correlate

**Comment by karalabe** [comment]
We've noticed a similar behavior when our goerli signer went out of disk. I think the last block only gets written partially to disk and than messes up things. We need to figure out what we can do to recover from this scenario (and shut geth down on OOD).

**Comment by jwasinger** [comment]
There is a WAL no?  Should be possible to recover from that upon startup.

**Comment by karalabe** [comment]
Managed to trace down the issue, will be fixed by #19544.

**Comment by karalabe** [comment]
Fixed on master.
